### PR TITLE
#9: Step 6 (closes #9)

### LINKED ISSUE #9: Step 6
## requirements
In this step will be getting rid of those pesky `~` 😃 We’ll be using wiro to skip the akka http DSL entirely. If you’re new to wiro, read https://blog.buildo.io/http-routes-at-buildo-1424250c41d3.

You can follow tutorial at https://buildo.github.io/wiro/ to get started.

### Tips

- the `Game` object should be split into a `GameApi` trait and a `GameApiImpl` class that implements that trait. This is probably redundant for this small app, but it makes sense in real world apps.
- wiro requires controller method to have the return a `Future[Either[ErrorType, A]]` where `ErrorType` and `A` are up to you. Since this application is not really asynchronous and it can’t really fail, I suggest you use `Throwable` as the error type and simply wrap the whole implementation into a `Future { ... }` block.
  - You will need to provide an `ToHttpResponse` for `Throwable`, which will never be used. Just to please the compiler add this before `deriveRouter` (we’re returning `null` but it doesn’t matter because it’s never going to be used).
  
       implicit def throwableResponse: ToHttpResponse[Throwable] = null

- in order for `Future { … }`  to work, it needs an implicit execution context (you already have it from the previous step. This ExecutionContext needs to come from the `Main`, so `GameApiImpl` will take it as a (possibly implicit) parameter. This is the stupidest form of Dependency Injection. It’s so simple and stupid that we love it 😃 
- One model you had to introduce in the previous step will become useless, can you guess which one?

### Solution
https://github.com/buildo/rps/compare/step-5...step-6

### PR DESCRIPTION
**Pull request by raasoft**
Closes #9

### tests performed

kapture 2018-07-13 at 15 47 42

#### Positive response 

Command:
> curl -d '{"userMove":"Rock"}' -H "Content-Type: application/json" -X POST http://localhost:8080/rps/play

Response: 
```sh
{"userMove":"Rock","computerMove":"Scissors","result":"Win"}%                                                                                                                                                                                  riccardo.ancona@Riccardos-MBP-2  ~/RAA/buildo/app-rpsosa   api-9-step_6 ⍟2  curl -d '{"userMove":"StopTHERock"}' -H "Content-Type: application/json" -X POST http://localhost:8080/rps/play
```

Command: 
> curl -d '{"userMove":"StopTHERock"}' -H "Content-Type: application/json" -X POST http://localhost:8080/rps/play

Response: 
``` sh
Failed decoding of '' of type 'Expected a value of type Move, but got StopTHERock'%
```

#### Negative response 

Command: 
> curl -d '{"userMove":"StopTHERock"}' -H "Content-Type: application/json" -X POST http://localhost:8080/rps/play

Response: 
``` sh
Failed decoding of '' of type 'Expected a value of type Move, but got StopTHERock'%
```

### CODE REVIEW

File: src/main/scala/rps/Game.scala [hunk 1]
```diff
@@ -4,24 +4,40 @@ import Move._
...
+trait GameApi {
+  @command
+  def play(
+      userMove: Move
+  ): Future[Either[Throwable, Response]]
+}
+
+// API implementation
+class GameApiImpl() extends GameApi {
+  import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (gabro) [inline src/main/scala/rps/Game.scala]: This works, but it's better to define a single `ExecutionContext` to use across all controllers, try passing the execution context to `GameApiImpl` as a parameter.

**Review by gabro — CHANGES_REQUESTED** [review]
Almost perfect 👍

File: src/main/scala/rps/Main.scala [hunk 2]
```diff
@@ -28,7 +29,9 @@ object RPSServer extends App with RouterDerivationModule {
 
   implicit def throwableResponse: ToHttpResponse[Throwable] = null
 
-  val gameRouter = deriveRouter[GameApi](new GameApiImpl)
+  val gameApiImpl = new GameApiImpl()(ec)
```
Comment (gabro) [inline src/main/scala/rps/Main.scala]: `ec` is implicit, you don't need to pass it explicitly ;)

File: src/main/scala/rps/Main.scala [hunk 3]
```diff
@@ -11,6 +11,7 @@ import akka.http.scaladsl.model.{
   HttpEntity
 }
 import wiro.Config
+import scala.concurrent.ExecutionContext._
```
Comment (gabro) [inline src/main/scala/rps/Main.scala]: This import seems unused

**Review by gabro — CHANGES_REQUESTED** [review]
Super minor things, then we can merge

File: src/main/scala/rps/Main.scala (hunk 2 above)
Comment (raasoft) [reply src/main/scala/rps/Main.scala]: I really need some tips on monday about implicit :3 Seems that there are a non-numerable set of ways where implicit is used, and this causes me lots of headaches. Apparently, implicit was used by Gödel too to demonstrate his incompleteness theorem.

Comment (gabro) [reply src/main/scala/rps/Main.scala]: lol.

Spoiler: there's only two ways of using implicits ;)

**Review by gabro — APPROVED** [review]
👍👌

**Pull request merged**